### PR TITLE
fix(rules): merge allowed subdirs correctly for Redirect_to_parent rules

### PR DESCRIPTION
**Pull request by anmonteiro**
- After https://github.com/ocaml/dune/pull/6637, the engine may redirect rule loading to the parent if it finds that it needs to find rules in the parent directory
- Part of generating rules in a directory is also finding out which sub-directories Dune is allowed to traverse; we store that in the record fields `build_dir_only_sub_dirs`
- However, while `Build_config.Rules.t` can now contain rules for multiple directories, `build_dir_only_sub_dirs` is still a flat set of sub-directories, leading Dune to (wrongly) assume that both directories which had their rules combined can traverse the same set of sub-directories, when that's not the case.
- This diff fixes that inconsistency by tracking the allowed set of sub-directories at each level
  - Specifically, this change unlocks traversing all directories when finding rules for recursive aliases, not just the directories in the source tree.

### CODE REVIEW

File: src/dune_engine/build_config.ml [hunk 1]
```diff
@@ -13,20 +13,25 @@ end
...
-    { build_dir_only_sub_dirs = Subdir_set.empty
+    { build_dir_only_sub_dirs = Path.Local.Map.empty
     ; directory_targets = Path.Build.Map.empty
     ; rules = Memo.return Rules.empty
     }
 
   let combine_exn r { build_dir_only_sub_dirs; directory_targets; rules } =
     { build_dir_only_sub_dirs =
-        Subdir_set.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
+        Path.Local.Map.merge r.build_dir_only_sub_dirs build_dir_only_sub_dirs
```
Comment (rgrinberg) [inline src/dune_engine/build_config.ml:29]: you can use `Map.union` here

File: src/dune_engine/load_rules.ml [hunk 2]
```diff
@@ -501,14 +501,19 @@ end = struct
...
-      { build_dir_only_sub_dirs : Subdir_set.t
+      { build_dir_only_sub_dirs : Subdir_set.t Path.Local.Map.t
       ; directory_targets : Loc.t Path.Build.Map.t
       ; rules : Rules.t Memo.Lazy.t
       }
 
     let combine_exn r { build_dir_only_sub_dirs; directory_targets; rules } =
       { build_dir_only_sub_dirs =
-          Subdir_set.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
+          Path.Local.Map.merge r.build_dir_only_sub_dirs build_dir_only_sub_dirs
```
Comment (rgrinberg) [inline src/dune_engine/load_rules.ml:511]: ditto

File: src/dune_rules/gen_rules.ml [hunk 3]
```diff
@@ -523,41 +541,53 @@ let gen_rules ~sctx ~dir components : Build_config.gen_rules_result Memo.t =
...
         match under_melange_emit_target with
-        | None -> Build_config.Rules build_config
-        | Some for_melange -> (
-          match gen_melange_emit_rules sctx ~dir for_melange with
-          | None -> Build_config.Redirect_to_parent build_config
-          | Some emit ->
-            Build_config.Rules
-              (Build_config.Rules.combine_exn build_config emit)))))
+        | None ->
+          let+ subdirs =
```
Comment (rgrinberg) [inline src/dune_rules/gen_rules.ml:563]: Do you mind moving this fix to a separate PR?

**Comment by anmonteiro** [comment]
Moved the automatic subdir stuff over to #7208 and fixed your review comments.

File: src/dune_engine/load_rules.ml [hunk 4]
```diff
@@ -660,6 +661,11 @@ end = struct
       Memo.return
         (Loaded.Build_under_directory_target { directory_target_ancestor })
     | Normal { rules; build_dir_only_sub_dirs; directory_targets } ->
+      let build_dir_only_sub_dirs =
+        Path.Local.Map.find build_dir_only_sub_dirs
+          (Path.Source.to_local sub_dir)
+        |> Option.value ~default:Subdir_set.empty
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply src/dune_engine/load_rules.ml:667]: I don't get how that would work if you're combining multiple `Build_config.Rules.t` where both come from different directories but have `Path.Local.root` keys.

[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_engine/load_rules.ml:667]: Roughly like this:

```ocaml
let combine (dir, map) (dir', map') =
  match Path.Build.descendant ~of_:dir dir' with
  | None -> (* impossible *)
  | Some p ->
    Path.Local.Map.foldi map' ~init:map ~f:(fun path dirs acc ->
      Path.Local.Map.add_exn acc (Path.Build.append_local p path) dirs)
```

[the remark this answers is not in the document]
Comment (anmonteiro) [reply src/dune_engine/load_rules.ml:667]: As I see it now, the choice comes to whether we want to pass `~dir` when 1. combining or when 2. creating the `Build_config.Rules.t` record.

[the remark this answers is not in the document]
Comment (anmonteiro) [reply src/dune_engine/load_rules.ml:667]: which one is that? I'd argue that my solution is simpler because it uses the directory we already have as the key in the map, rather than piecing it together afterwards (and requiring the user of the API to pass the extra directory arg)

[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_engine/load_rules.ml:667]: > which one is that? I'd argue that my solution is simpler because it uses the directory we already have as the key in the map

The problem is that this allows invalid keys in the map. So it would need an extra validation step.

> and requiring the user of the API to pass the extra directory arg)

Why should the user of the API pass the extra directory arg? When we evaluate the user callback we know exactly what the directory is.

[the remark this answers is not in the document]
Comment (anmonteiro) [reply src/dune_engine/load_rules.ml:667]: I think I understand. What paths are wrong in my case? The subdirectories that don't end up containing rules?

[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_engine/load_rules.ml:667]: In your case? You mean in how you use the API? I don't think there's anything wrong, it seems like you're using it correctly. What I mean is that it's possible to use this API is correctly, for example by passing `Path.Local.root` as the key from any other directory. While with my suggestion, `Path.Local.root` corresponds to the current directory and hence it's only possible to provide directories under it, as expected.

[the remark this answers is not in the document]
Comment (anmonteiro) [reply src/dune_engine/load_rules.ml:684]: I changed this PR to your suggestion, and I feel like it's a lot easier to follow what's happening when `Gen_rules` needs only to worry about looking up sub-directories for `Path.Local.root`.

I pushed it as a separate commit so that we can still have a chance to compare the 2 alternatives ([this commit](https://github.com/ocaml/dune/pull/7207/commits/a21a577428e292cf96a6be32df875c4a705dd2c4) vs [entire diff](https://github.com/ocaml/dune/pull/7207/files))

**Comment by anmonteiro** [comment]
@snowleopard I requested your review on this one since it touches the engine and you were a reviewer of #6637.

File: src/dune_engine/load_rules.ml [hunk 5]
```diff
@@ -501,14 +501,15 @@ end = struct
...
+      { build_dir_only_sub_dirs : Subdir_set.t Path.Build.Map.t
       ; directory_targets : Loc.t Path.Build.Map.t
       ; rules : Rules.t Memo.Lazy.t
       }
 
     let combine_exn r { build_dir_only_sub_dirs; directory_targets; rules } =
       { build_dir_only_sub_dirs =
-          Subdir_set.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
+          Path.Build.Map.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
+            ~f:(fun _ a b -> Some (Subdir_set.union a b))
```
Comment (snowleopard) [inline src/dune_engine/load_rules.ml:512]: Why is this code written twice? Please factor it out.

File: src/dune_rules/gen_rules.ml [hunk 6]
```diff
@@ -457,49 +465,50 @@ let gen_melange_emit_rules sctx ~dir ({ stanza_dir; stanza } as for_melange) =
                (Path.Source.to_string parent_melange_emit_dir)
            ]))
 
-let empty_rules =
+let empty_rules ~dir =
   { Build_config.Rules.build_dir_only_sub_dirs =
-      Subdir_set.These (String.Set.of_keys automatic_sub_dirs_map)
+      Path.Build.Map.singleton dir
+        (Subdir_set.These (automatic_subdirs (Path.Build.explode dir)))
```
Comment (snowleopard) [inline src/dune_rules/gen_rules.ml:471]: Another duplicated code snippet makes me think that there is some module to be written where these helper functions are going to live.

File: src/dune_engine/load_rules.ml (hunk 5 above)
Comment (anmonteiro) [reply src/dune_engine/load_rules.ml]: Thanks, added a new module in [522eaea](https://github.com/ocaml/dune/pull/7207/commits/522eaead7b8b3c01d3a67573fc0d4fb346a568be) with this logic. I think it becomes more explicit.

File: src/dune_rules/gen_rules.ml (hunk 6 above)
Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: took care of this too in [522eaea](https://github.com/ocaml/dune/pull/7207/commits/522eaead7b8b3c01d3a67573fc0d4fb346a568be)

File: src/dune_engine/load_rules.ml (hunk 5 above)
Comment (snowleopard) [reply src/dune_engine/load_rules.ml]: Thanks! Agreed that it's better now.

File: src/dune_engine/load_rules.ml [hunk 7]
```diff
@@ -660,6 +661,10 @@ end = struct
       Memo.return
         (Loaded.Build_under_directory_target { directory_target_ancestor })
     | Normal { rules; build_dir_only_sub_dirs; directory_targets } ->
+      let build_dir_only_sub_dirs =
+        Path.Build.Map.find build_dir_only_sub_dirs dir
```
Comment (snowleopard) [inline src/dune_engine/load_rules.ml:665]: It probably makes sense to move this bit of logic to `Build_only_sub_dirs`. If we move a couple other helper functions there (like `singleton`) we can probably make the type abstract, which seems like a win in the long term.

Comment (anmonteiro) [reply src/dune_engine/load_rules.ml]: I like that idea. Did that in [cf48a7a](https://github.com/ocaml/dune/pull/7207/commits/cf48a7a982e6b427312f610b6b3598c6cbf218c9)

Comment (snowleopard) [reply src/dune_engine/load_rules.ml]: Nice!

File: src/dune_engine/build_config.mli [hunk 8]
```diff
@@ -14,13 +14,27 @@ end
...
+  module Build_only_sub_dirs : sig
+    (** The set of either directory targets or internally generated directories,
+        indexed by their parent build directory. *)
+    type t
+
+    val empty : t
+
+    val singleton : dir:Path.Build.t -> Subdir_set.t -> t
+
+    val find : t -> Path.Build.t -> Subdir_set.t option
```
Comment (snowleopard) [inline src/dune_engine/build_config.mli:31]: I feel like there isn't any meaningful distinction between `None` and `Some Subdir_set.empty` outcomes, which suggests that we can get rid of the `option` here.

Comment (anmonteiro) [reply src/dune_engine/build_config.mli]: [c000eb5](https://github.com/ocaml/dune/pull/7207/commits/c000eb5cd7eb19c88193d7f6c387edd55e7cad4b)

File: src/dune_rules/gen_rules.ml [hunk 9]
```diff
@@ -350,6 +350,13 @@ let automatic_sub_dirs_map =
     ; (Artifacts.Bin.bin_dir_basename, Bin)
     ]
 
+let automatic_subdirs path_components =
+  match List.last path_components with
```
Comment (snowleopard) [inline src/dune_rules/gen_rules.ml:354]: It's odd to first `explode` a directory into path components, thus constructing the whole list, and then immediately throw away all elements but the last one. How about we fuse the two steps?

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: Used a slightly different approach in [e251d39](https://github.com/ocaml/dune/pull/7207/commits/e251d39b278b5de0f506b40920985f5464b7e828) that doesn't have these problems.

File: src/dune_rules/gen_rules.ml [hunk 10]
```diff
@@ -350,6 +350,13 @@ let automatic_sub_dirs_map =
     ; (Artifacts.Bin.bin_dir_basename, Bin)
     ]
 
+let automatic_subdirs dir =
+  match Path.Build.basename_opt dir with
+  | None -> String.Set.of_keys automatic_sub_dirs_map
+  | Some comp ->
+    if String.Map.mem automatic_sub_dirs_map comp then String.Set.empty
```
Comment (snowleopard) [inline src/dune_rules/gen_rules.ml:357]:
```suggestion
  | Some basename ->
    if String.Map.mem automatic_sub_dirs_map basename then String.Set.empty
```

**Review by snowleopard — APPROVED** [review]
Thanks for addressing all my comments. Looks good now!

**Comment by anmonteiro** [comment]
Thanks!

**Comment by anmonteiro** [comment]
Squashed my commits.

File: src/dune_rules/gen_rules.ml (hunk 9 above)
Comment (rgrinberg) [reply src/dune_rules/gen_rules.ml]: > It's odd to first explode a directory into path components, thus constructing the whole list, and then immediately throw away all elements but the last one. How about we fuse the two steps?

Note that we aren't exploding the directory into path components in the rules. It's already exploded for us by the engine. So if the last component of the path is already provided for us, we should use it.

File: src/dune_engine/load_rules.ml [hunk 11]
```diff
@@ -501,14 +501,15 @@ end = struct
 
   module Normal = struct
     type t =
-      { build_dir_only_sub_dirs : Subdir_set.t
+      { build_dir_only_sub_dirs : Build_config.Rules.Build_only_sub_dirs.t
```
Comment (rgrinberg) [inline src/dune_engine/load_rules.ml:504]: In `make_rules_gen_result` we verify that both `rules` and `directory_targets` only contain valid keys. We should similar validation for this map as well.

Comment (anmonteiro) [reply src/dune_engine/load_rules.ml:504]: did that in [72bea81](https://github.com/ocaml/dune/pull/7207/commits/72bea812588daf8a2c34c1a5ee2d566cfaa3c5ab)

File: src/dune_rules/gen_rules.ml (hunk 9 above)
Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: @rgrinberg tried something like what you described in [04e678c](https://github.com/ocaml/dune/pull/7207/commits/04e678cf33c02caca8a87e3daf2a8e6f8b8922fe), though it's a bit more invasive. Which version do you prefer?

File: src/dune_rules/gen_rules.ml [hunk 12]
```diff
@@ -405,29 +413,34 @@ let rec under_melange_emit_target ~dir =
       | None -> under_melange_emit_target ~dir:parent
       | Some stanza -> Memo.return @@ Some { stanza_dir = parent; stanza }))
 
-let melange_emit_rules sctx { stanza_dir; stanza } =
+let rules_for ~dir components rules =
```
Comment (rgrinberg) [inline src/dune_rules/gen_rules.ml:416]: I find the way this is factored is making less clear. In some functions, we pass `components`, elsewhere, we use this `components` to compute the `automatic_subdirs_map`. How about we just compute `automatic_subdirs_map` right away and pass that around.

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: check [33a7336](https://github.com/ocaml/dune/pull/7207/commits/33a7336ae9877706f66e693382a96b98a2384933) where I implemented your suggestion.

File: src/dune_rules/gen_rules.ml (hunk 9 above)
Comment (snowleopard) [reply src/dune_rules/gen_rules.ml]: > > It's odd to first explode a directory into path components, thus constructing the whole list, and then immediately throw away all elements but the last one. How about we fuse the two steps?
> 
> Note that we aren't exploding the directory into path components in the rules. It's already exploded for us by the engine. So if the last component of the path is already provided for us, we should use it.

@rgrinberg I'm not quite sure what you mean. In the version of the code which I was commenting on, we had an explicit `explode` followed by `List.last`. If we could get an exploded list from the caller, it would certainly be better but still, if we ever only look at the last component, the intermediate list of components is unnecessary (and maybe we need to change some API to avoid generating intermediate lists).

Comment (rgrinberg) [reply src/dune_rules/gen_rules.ml]: There must have been an extra explode that I didn't notice. It's gone now I suppose.

>  if we ever only look at the last component, the intermediate list of components is unnecessary (and maybe we need to change some API to avoid generating intermediate lists).

We use the components elsewhere.

> (and maybe we need to change some API to avoid generating intermediate lists).

That would make sense. I don't see why the build engine has to pre-emptively split the path.
